File: mergeSort/mergeSort.cpp
#include <iostream>
#include <vector>
#include <math.h>
#include <chrono>
#include <random>

std::vector<int> mergeSort_Helper(std::vector<int> left, std::vector<int> right);

std::vector<int> bubbleSort(std::vector<int>& unsortedArray)
{
    int lengthOfArray = unsortedArray.size();
    for (int i = 0; i < lengthOfArray; i++)
    {
        for (int j = 0; j < lengthOfArray - i - 1; j++)
        {
            if (unsortedArray[j] > unsortedArray[j + 1])
            {
                int greaterVal = unsortedArray[j];
                int lowerVal = unsortedArray[j + 1];
                unsortedArray[j] = lowerVal;
                unsortedArray[j + 1] = greaterVal;
            }
        }
    }
    return unsortedArray;
}

std::vector<int> mergeSort(std::vector<int>& unsortedArray)
{
    if (unsortedArray.size() <= 1)
    {
        return unsortedArray;
    }

    size_t mid = unsortedArray.size()/2;
    std::vector<int> left_array;
    std::vector<int> right_array;
    for (size_t i = 0; i < mid; ++i)
    {
        left_array.push_back(unsortedArray[i]);
    }
    for (size_t i = mid; i < unsortedArray.size(); i++)
    {
        right_array.push_back(unsortedArray[i]);
    }

    left_array = mergeSort(left_array);
    right_array = mergeSort(right_array);
    
    return mergeSort_Helper(left_array, right_array);
}

std::vector<int> mergeSort_Helper(std::vector<int> left, std::vector<int> right)
{
    std::vector<int> result;
    size_t i = 0;
    size_t j = 0;

    while (i < left.size() && j < right.size())
    {
        if(left[i] < right[j])
        {
            result.push_back(left[i]);
            i++;
        }
        else
        {
            result.push_back(right[j]);
            j++;
        }
    }

    while (i < left.size())
    {
        result.push_back(left[i]);
        i++;
    }
    while (j < right.size())
    {
        result.push_back(right[j]);
        j++;
    }

    return result;

}


int main()
{


    //set up needed vars
    double min;
    double max;
    double n;
    std::vector<int> unsortedArray;

    //get info from user
    std::cout << "Min: ";
    std::cin >> min;
    std::cout << "Max: ";
    std::cin >> max;
    std::cout << "Number Of Elements: ";
    std::cin >> n;

    //randomly generate array given params
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> distrib(min, max);

    for (int i = 0; i < n; i++)
    {
        unsortedArray.push_back(distrib(gen));
    }

    //time array and actuall call the sorting function
    auto startTime = std::chrono::high_resolution_clock::now();
    std::vector<int> sortedArray = mergeSort(unsortedArray);
    auto endTime = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> totalTime = endTime - startTime;

    //print out the array and the time it took to sort it
    //for (int i = 0; i < sortedArray.size(); i++)
    //{
    //    std::cout << sortedArray[i] << ", ";
    //}
    std::cout << "\n\nTotal Time To Sort Array: " << totalTime.count() << " seconds\n";


    return 0;
}

File: mergeSort/mergeSort.js
unsortedList = [50, 11, 23, 3, 38, 39, 0, 7, 0, 15, 6, 9, 16, 25, 30, 41, 30, 29, 23, 39, 20, 18, 20, 20, 5, 13, 5, 9, 10, 6]

function mergeSort(unsortedList)
{
    if(unsortedList.length <= 1)
    {
        return unsortedList;
    }

    let mid = Math.floor(unsortedList.length / 2);
    let leftArray = unsortedList.slice(0, mid);
    let rightArray = unsortedList.slice(mid);

    leftArray= mergeSort(leftArray);
    rightArray = mergeSort(rightArray);

    return mergeSort_Helper(leftArray, rightArray);
}

function mergeSort_Helper(left, right)
{
    let result = []
    let i = 0;
    let j = 0;

    while(i < left.length && j < right.length)
    {
        if(left[i] < right[j])
        {
            result.push(left[i]);
            i++;
        }
        else
        {
            result.push(right[j]);
            j++;
        }
    }

    while(i < left.length)
    {
        result.push(left[i]);
        i++;
    }
    while(j < right.length)
    {
        result.push(right[j]);
        j++
    }

    return result;
}
